src/Orleans/Streams/Internal/StreamConsumerExtension.cs

```diff
@@ -63,50 +67,90 @@ internal StreamSubscriptionHandleImpl<T> SetObserver<T>(GuidId subscriptionId, S
             }
         }
 
-        internal bool RemoveObserver(GuidId subscriptionId)
+        public Task OnSubscriptionChange<T>(Func<StreamSubscriptionHandle<T>, Task> onAdd, Func<string, IStreamIdentity, Guid, Task> onRemove = null)
+        {
+            this.onSubscriptionAddAction = onAdd;
```

Comment:
thanks for the suggestion ! I do like the ISubscriptionChangeHandler for handling the first question.  I will 

not sure how is it going to solve the reflection issue yet. But I assume it is something like what we did for StreamSubscriptionHandler<T> and IStreamSubscriptionHandler. Will try prototype it